src/az_ulib_ipc/az_ulib_ipc.c

```diff
@@ -0,0 +1,352 @@
+      result = AZ_ULIB_NO_SUCH_ELEMENT_ERROR;
+      /*az_ulib_ipc_try_get_interface_with_max_interface_instances_failed*/
+    } else if ((result = get_instance(ipc_interface)) == AZ_ULIB_SUCCESS) {
+      /*az_ulib_ipc_try_get_interface_succeed*/
+      *interface_handle = ipc_interface;
+    }
+  }
+  az_pal_os_lock_release(&(ipc->lock));
+
+  return result;
+}
+
+AZ_ULIB_RESULT
+_az_ulib_ipc_try_get_interface(
+    const char *const name,
+    az_ulib_version version,
+    AZ_ULIB_VERSION_MATCH_CRITERIA match_criteria,
+    _az_ulib_ipc_interface_handle *interface_handle) {
+  AZ_UCONTRACT(
+      /*az_ulib_ipc_try_get_interface_with_ipc_not_initialized_failed*/
+      AZ_UCONTRACT_REQUIRE_NOT_NULL(ipc, AZ_ULIB_NOT_INITIALIZED_ERROR),
+      /*az_ulib_ipc_try_get_interface_with_null_name_failed*/
+      AZ_UCONTRACT_REQUIRE_NOT_NULL(name, AZ_ULIB_ILLEGAL_ARGUMENT_ERROR),
+      /*az_ulib_ipc_try_get_interface_with_null_handle_failed*/
+      AZ_UCONTRACT_REQUIRE_NOT_NULL(interface_handle, AZ_ULIB_ILLEGAL_ARGUMENT_ERROR));
+  return _az_ulib_ipc_try_get_interface_no_test(name, version, match_criteria, interface_handle);
+}
+
+AZ_ULIB_RESULT
+_az_ulib_ipc_get_interface_no_test(
+    _az_ulib_ipc_interface_handle original_interface_handle,
+    _az_ulib_ipc_interface_handle *interface_handle) {
+  AZ_ULIB_RESULT result;
+
+  az_pal_os_lock_acquire(&(ipc->lock));
+  {
+    _az_ulib_ipc_interface *ipc_interface = original_interface_handle;
+    if (ipc_interface->interface_descriptor == NULL) {
+      /*az_ulib_ipc_get_interface_with_unpublished_interface_failed*/
+      result = AZ_ULIB_NO_SUCH_ELEMENT_ERROR;
+      /*az_ulib_ipc_get_interface_with_max_interface_instances_failed*/
+    } else if ((result = get_instance(ipc_interface)) == AZ_ULIB_SUCCESS) {
+      /*az_ulib_ipc_get_interface_succeed*/
+      *interface_handle = ipc_interface;
+    }
+  }
+  az_pal_os_lock_release(&(ipc->lock));
+
+  return result;
+}
+
+AZ_ULIB_RESULT
+_az_ulib_ipc_get_interface(
+    _az_ulib_ipc_interface_handle original_interface_handle,
+    _az_ulib_ipc_interface_handle *interface_handle) {
+  AZ_UCONTRACT(
+      /*az_ulib_ipc_get_interface_with_ipc_not_initialized_failed*/
+      AZ_UCONTRACT_REQUIRE_NOT_NULL(ipc, AZ_ULIB_NOT_INITIALIZED_ERROR),
+      /*az_ulib_ipc_get_interface_with_null_original_interface_handle_failed*/
+      AZ_UCONTRACT_REQUIRE_NOT_NULL(original_interface_handle, AZ_ULIB_ILLEGAL_ARGUMENT_ERROR),
+      /*az_ulib_ipc_get_interface_with_null_interface_handle_failed*/
+      AZ_UCONTRACT_REQUIRE_NOT_NULL(interface_handle, AZ_ULIB_ILLEGAL_ARGUMENT_ERROR));
+  return _az_ulib_ipc_get_interface_no_test(original_interface_handle, interface_handle);
+}
+
+AZ_ULIB_RESULT
+_az_ulib_ipc_release_interface_no_test(_az_ulib_ipc_interface_handle interface_handle) {
+  _az_ulib_ipc_interface *ipc_interface = (_az_ulib_ipc_interface *)interface_handle;
+  AZ_ULIB_RESULT result;
+
+  if (ipc_interface->ref_count == 0) {
```

Comment:
>if (ipc_interface->ref_count == 0) { [](start = 2, length = 36)

lock around block for check and atomic dec

Reply:
good catch! Thanks!

---
In reply to: [338324202](https://github.com/Azure/azure-ulib-c/pull/40#discussion_r338324202) [](ancestors = 338324202)